XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+
+Called immediately after the return callback is completed. If the callback
+threw but was caught by a domain or `uncaughtException` `didThrow` will be set
+to `true`. If the callback threw but was not caught then the process will exit
+immediately without calling `post()`.
+
+#### `destroy(id)`
+
+* `id` {Number}
+
+Called either when the class destructor is run, or if the resource is marked as
+free. The destructor will usually run when explicitly called (the case for
+handles) or when a request has completed. In at least one case this will be
+triggered by GC. In the case of shared or cached resources, such as
+`HTTPParser`, `destroy()` will be manually called when the TCP connection no
+longer needs it.
+
+Because the callback is called during deconstruction of the class instance it
+is not safe to access the JS handle during `destroy()` execution. Reason this
+callback is important is to allow resource cleanup. While APIs like `WeakMap()`
+are great, they are currently not iterable and not as performant. Taking an
+above example, here is how `destroy` can be used for resource cleanup:
+
+```js
+const handleMap = new Map();
+
+function HandleInst(inst) {
+  // Emulate performance.now()
+  const t = process.hrtime();
+  this.time = t[0] * 1e3 + t[1] / 1e6;
+  this.inst = inst;
+}
+
+function init(id) {
+  handleMap.set(id, new HandleInst(this));
+}
+
+function destroy(id) {
+  handleMap.delete(id);
+}
+```
+
+
+## API Exceptions
+
+### net Client connection Event
+
+Technically the `pre()`/`post()` events of the `'connection'` event for
+`net.Server` would place the server as the active id. Problem is that this is
+not intuitive in how the asynchronous chain would propagate. So instead make
+the client the active id for the duration of the `'connection'` callback.
+
+### Reused Resources
+
+Resources like `HTTPParser` are reused throughout the lifetime of the
+application. This means node will have to synthesize the `init()` and
+`destroy()` calls. Also the id on the class instance will need to be changed
+every time the resource is acquired for use.
+
+For shared resources like `TimerWrap` this is not necessary since there is a
+unique JS handle that will contain the unique id necessary for the calls.
+
+
+## Notes
+
+### `process.nextTick()` and Promises
+
+Execution of the `nextTickQueue` and `MicrotaskQueue` are slightly special
+asynchronous cases. Because they execute in the same `MakeCallback()` as the
+asynchronous callback they will ultimately end up with the same parent as the
+originating call.
+
+In the case of `nextTick()`, many calls are meant to call or queue other tasks.
+Which would easily end up causing async wrap to incur more of a cost than
+calling each callback in the `nextTickQueue`. Though not calling the hooks
+would lead to loss of stack information.
+
+Node currently doesn't have sufficient API to notify calls to Promise
+callbacks. In order to do so node would have to override the native
+implementation.
```

Comment:
@chrisdickinson If we absolutely have to I'd prefer shimming native promises over not shipping this. We don't currently offer any type of support for Promises in this way. e.g. domains loosing context:
```js
const domain = require('domain');
const print = process._rawDebug;
domain.create().run(() => Promise.resolve(1).then((val) => print(process.domain)));
// output: undefined
```
And TBH refusing this type of insight into node just because v8 doesn't provide us with more than a black box is a serious pain.

If, in fact, the CTC more unanimously decides that it's important that we wait for v8 to provide the appropriate APIs to accomplish this then I'd at least like to get approval for the EP so I can finish the implementation exposed via `process.binding()`. Then at least it's exposed in a way that can still be useful to many and will be ready once v8 provides what we need.